perf/tests: Define an OMB perf template

bharathv opened the pull request:
## Cover letter

Adds a ducktape perf template that runs a standard
OMB(OpenMessagingBenchmark) workload.

This patch restructures the ducktape OMB code by
templatizing the input workload and driver files.
This makes it easy to add any new drivers/workloads
and run them as a part of the test.

The end goal is to define a realistic benchmark that
can catch perf regressions by fine tuning the workloads
and this patch lays the foundation for that.

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #5140

## Release notes

* none

Otherwise, add one or more of the following sections. A section must have
at least 1 bullet point. You can add multiple sections with multiple
bullet points if this PR represents multiple release note items. See
the CONTRIBUTING.md guidelines for more details.

### Features

* Short description of the feature. Explain how to configure the new feature if applicable.

### Improvements

* Short description of how this PR improves redpanda.

-->

rystsov commented on tests/rptest/services/openmessaging_benchmark.py:
```diff
@@ -118,77 +119,60 @@ class OpenMessagingBenchmark(Service):
+        """Default configuration to use based on the environment in which the driver is running."""
+        # We pick different defaults depending on the environment we are in. Dedicated nodes get a heavier
+        # workload by default if sample is picked is picked.
         if self.redpanda.dedicated_nodes:
```
This looks weird. I understand previously `dedicated_nodes` used to adjust duration of the tests. So the existing 4 tests:

- test_multiple_subscriptions_omb
- test_multiple_producers_consumers_omb
- test_multiple_topics_omb
- test_default_omb_configuration

were adjusted according to the environment.

Not OpenMessagingBenchmark is parametrized and dedicated_nodes replaces the whole template. It's like as in the previous version `test_multiple_topics_omb` was silently replaced with `test_default_omb_configuration`. It's very implicit and may lead to unexpected behavior.

For example let's took this test

```
@matrix(driver_idx=[0, 1], workload_idx=[0])
def test_default_omb_configuration(self, driver_idx, workload_idx)
```

The matrix parameter expands into `[[0,0],[1,0]]` but `dedicated_nodes` is true it is being replaced with `[[0,0],[0,0]]` and we execute the same test twice withing it does something different while it isn't true.

It may be better if we have solved this problem at higher level. E.g. we could have

  - openmessaging_benchmark_cdt_test.py
  - openmessaging_benchmark_local_test.py
  - openmessaging_benchmark_etc_test.py

files and just include / exclude them for different suites

rystsov commented on tests/rptest/perf/openmessaging_perf_test.py:
```diff
@@ -0,0 +1,44 @@
+                                                num_brokers=3)
+
+    @cluster(num_nodes=9)
+    @parametrize(driver_idx=3, workload_idx=2)
```
Let's use strings instead of indexes: it's easier to search (ctrl-f or `git grep`) and it's easier to understand what they do. Compare 

```
@cluster(num_nodes=9)
@parametrize(driver_idx=3, workload_idx=2)
def test_perf_with_idempotence_and_max_in_flight(...
```

with

```
@cluster(num_nodes=9)
@parametrize(
  driver="ACK_ALL_GROUP_LINGER_1MS_IDEM_MAX_IN_FLIGHT",
  workload="TOPIC1_PART100_1KB_4PROD_1250K_RATE")
def test_perf_with_idempotence_and_max_in_flight(...
```

bharathv commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/371514100e2652ac7b54114973ed1e364a7b6c85..3ccda2a7474f4525b9c5f489633cb4a9af283ae2) addresses Denis' first round of comments and rebases on top of dev tip.

rystsov commented:
The changes looks good but we need to embed smoke check into the ducktape tests to detect when latency falls out of the accepted range to fail the test.

bharathv commented on tests/docker/Dockerfile:
```diff
@@ -142,8 +142,8 @@ RUN python3 -m pip install --upgrade --force pip && \
 # Install the OMB tool
-RUN git -C /opt clone https://github.com/redpanda-data/openmessaging-benchmark.git && \
-    cd /opt/openmessaging-benchmark && git reset --hard cde313756faa1ebd0114a35d6fc3b09198559738 && mvn package
+RUN git -C /opt clone https://github.com/bharathv/openmessaging-benchmark.git && \
```
This is a temp change to showcase the e2e patch. Had to make some changes in the way we chart the metrics. You can find the patch [here](https://github.com/bharathv/openmessaging-benchmark/commit/aa943b284bf81796a968a87e788f9b2c0747de10), will update to official repos before merging.

ZeDRoman replied:
Don't forget to change it also in CDT setup
https://github.com/redpanda-data/vtools/blob/dev/qa/image/packer/install-deps.sh#L195

bharathv reviewed (commented):
@rystsov Added validation rules logic that runs on the metrics data. We should revisit the actual rules once we establish historical baselines. Until then they are pretty basic.

ZeDRoman commented on tests/rptest/services/openmessaging_benchmark_configs.py:
```diff
@@ -0,0 +1,170 @@
+        "topics": 1,
+        "partitions_per_topic": 100,
+        "subscriptions_per_topic": 1,
+        "consumer_per_subscription": 16,
```
How it works with only 2 worker instances?
I thought that every consumer needs separate worker

ZeDRoman commented on tests/rptest/services/openmessaging_benchmark_metrics_validator.py:
```diff
@@ -0,0 +1,61 @@
+        return (lambda x: x >= min,
+                f"Expected to be >= {min_val}, check failed.")
+
+    VALIDATORS = {
```
How you determined values here?

ZeDRoman reviewed (commented):
It looks great

ZeDRoman commented on tests/rptest/services/openmessaging_benchmark.py:
```diff
@@ -16,12 +16,14 @@
 class OpenMessagingBenchmarkWorkers(Service):
-    PERSISTENT_ROOT = "/var/lib/openmessaging"
-    STDOUT_STDERR_CAPTURE = os.path.join(PERSISTENT_ROOT, "worker.log")
+    ARTIFACTS_DIR = "/var/lib/openmessaging"
```
I think that we should use standard names everywhere

bharathv commented:
Latest [force-push](https://github.com/redpanda-data/redpanda/compare/532d25e470ea6530a0ed38b878add2920c21dd2f..2344b7386e1c35a52e4d4f8b74c61bca17170110) addresses all the comments and updates the OMB commit with relevant charting changes.

This has been lying around for too long, can someone signoff on this if there are no additional comments? Thanks. 
(Just for the context, this change is a pre-req for automated nightly OMB run)

ZeDRoman commented on tests/rptest/perf/openmessaging_perf_test.py:
```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Vectorized, Inc.
```
2022

ZeDRoman commented on tests/rptest/services/openmessaging_benchmark_configs.py:
```diff
@@ -0,0 +1,236 @@
+    E2E_LATENCY_MAX = "aggregatedEndToEndLatencyMax"
+    AVG_THROUGHPUT_MBPS = "throughputMBps"
+
+    def range(min_val, max_val):
```
Maybe we should move these validators to other file?
